source/scheduler.hpp

```diff
@@ -1,46 +1,71 @@
 #pragma once
 
-#include <memory>
 #include <queue>
 
 #include "event.hpp"
 
 namespace sim {
 
-struct EventComparator {
-    bool operator()(const std::unique_ptr<Event>& lhs,
-                    const std::unique_ptr<Event>& rhs) const {
-        return (*lhs.get()) > (*rhs.get());
-    }
-};
-
 // Scheduler is implemented as a Singleton class
 // which provides a global access to a single instance
-class Scheduler {
+template <typename TEvent>
+requires requires(const TEvent& e) {
+    { e.get_time() } -> std::same_as<Time>;
+}
+class SchedulerTemplate {
+    static_assert(
+        std::is_constructible_v<bool, decltype(std::declval<TEvent>() >
+                                               std::declval<TEvent>())>,
+        "Event must have operator <");
+    static_assert(std::is_invocable_v<TEvent>, "Event must be invocable");
+
 public:
     // Static method to get the instance
-    static Scheduler& get_instance() {
-        static Scheduler instance;
+    static SchedulerTemplate& get_instance() {
+        static SchedulerTemplate instance;
         return instance;
     }
 
-    void add(std::unique_ptr<Event> event);
-    void clear();  // Clear all events
-    bool tick();
-    Time get_current_time();
+    void add(const TEvent& event) { m_events.emplace(event); }
+
+    // Clear all events
+    void clear() {
+        m_events =
+            std::priority_queue<TEvent, std::vector<TEvent>, EventComparator>();
+    }
+
+    bool tick() {
+        if (m_events.empty()) {
+            return false;
+        }
+
+        TEvent event = m_events.top();
+        m_events.pop();
+        m_current_event_local_time = event.get_time(); 
+        event.operator()();
+        return true;
+    }
+
+    Time get_current_time() { return m_current_event_local_time; }
 
 private:
+    struct EventComparator {
+        bool operator()(const TEvent& lhs, const TEvent& rhs) const {
+            return lhs > rhs;
+        }
+    };
+
     // Private constructor to prevent instantiation
-    Scheduler() {}
+    SchedulerTemplate() {}
```

Comment:
Initialize the fields:
```suggestion
    SchedulerTemplate() : m_events(), m_current_event_local_time(0) {}
```